godel_scan_analysis/src/scan_roughness_scoring.cpp

```diff
@@ -0,0 +1,133 @@
+#include "godel_scan_analysis/scan_roughness_scoring.h"
+
+#include "godel_scan_analysis/scan_algorithms.h"
+
+#include <ros/ros.h> // debug
+
+#include <math.h> // isfinite
+
+/*
+  I have a variety of ideas for making this code faster if it was ever needed:
+  1) Keep scan data as an array of x and array of y for better cache coherency on windowing
+  2) The windowing could be made to operate in O(n) - touching every point only once
+  3) Pre-calculate x values (which are known)
+  4) Pre-calculate colors
+  5) Perhaps return an index map like PCL does from the filter function so
+     we don't need to make a copy of the scan
+  6) Pre-allocate or use stack space (std::array?) for storage space
+*/
+
+namespace
+{
+  typedef godel_scan_analysis::RoughnessScorer::Cloud Cloud;
+  typedef godel_scan_analysis::RoughnessScorer::ColorCloud ColorCloud;
+
+  const static unsigned WINDOW_SIZE = 30;
+
+  // Preprocess clouds
+  static rms::Scan<double> filterCloudAndBuildScan(const Cloud& in)
+  {
+    rms::Scan<double> scan;
+    scan.points.reserve(in.points.size());
+    for (std::size_t i = 0; i < in.points.size(); ++i)
+    {
+      rms::Point<double> pt;
+      pt.x = in.points[i].x;
+      pt.y = in.points[i].z;
+
+      if (std::isfinite(pt.y)) scan.points.push_back(pt);
+    }
+
+    return scan;
+  }
+
+  static inline double constrainValue(double min, double max, double val)
+  {
+    return (val > max) ? max : ((val < min) ? min : val); 
+  }
+
+  // Takes one point and makes a colored pcl point from it
+  static pcl::PointXYZRGB makeColoredPoint(const rms::Point<double>& pt, double score)
+  {
+    // TODO: put these colorization values into the params struct
+    static const double max_score = 0.00003;
+    static const double min_score = 0.0;
```

Comment:
Absolutely. As the comment says, these should eventually be set via some kind of call, but for now constant is good.